src/parser/syntax.rs

```diff
@@ -0,0 +1,745 @@
+    move |from: Span<'_>| {
+        let (_, is_eof) = opt(nom::combinator::eof)(from)?;
+        if is_eof.is_some() {
+            return Ok((from, None));
+        }
+        let (next, meta) = opt(char('!'))(from)?;
+        if meta.is_some() {
+            let (end, syntax) = parse_syntax()(next)?;
+            Ok((end, Some((true, syntax))))
+        } else {
+            let (end, syntax) = parse_syntax()(from)?;
+            Ok((end, Some((false, syntax))))
+        }
+    }
+}
+
+#[cfg(test)]
+pub mod tests {
+    use crate::symbol::LurkSym;
+    use blstrs::Scalar;
+    use nom::Parser;
+    #[cfg(not(target_arch = "wasm32"))]
+    use proptest::prelude::*;
+
+    use super::*;
+    #[allow(unused_imports)]
+    use crate::{char, keyword, list, num, str, symbol, uint};
+
+    fn test<'a, P>(mut p: P, i: &'a str, expected: Option<Syntax<Scalar>>) -> bool
+    where
+        P: Parser<Span<'a>, Syntax<Scalar>, ParseError<Span<'a>, Scalar>>,
+    {
+        match (expected, p.parse(Span::<'a>::new(i))) {
+            (Some(expected), Ok((_i, x))) if x == expected => true,
+            (Some(expected), Ok((_i, x))) => {
+                println!("input: {:?}", i);
+                println!("expected: {} {:?}", expected.clone(), expected);
+                println!("detected: {} {:?}", x.clone(), x);
+                false
+            }
+            (Some(..), Err(e)) => {
+                println!("{}", e);
+                false
+            }
+            (None, Ok((i, x))) => {
+                println!("input: {:?}", i);
+                println!("expected parse error");
+                println!("detected: {:?}", x);
+                false
+            }
+            (None, Err(_e)) => true,
+        }
+    }
+
+    #[test]
+    fn unit_parse_string() {
+        assert!(test(parse_string(), "\"foo\"", Some(str!("foo"))));
+        assert!(test(parse_string(), "\"fo\\no\"", Some(str!("fo\no"))));
+        assert!(test(
+            parse_string(),
+            "\"fo\\u{00}o\"",
+            Some(str!("fo\u{00}o"))
+        ));
+        assert!(test(parse_string(), "\"foo\\   \"", Some(str!("foo"))));
+    }
+
+    #[test]
+    fn unit_parse_symbol() {
+        assert!(test(parse_symbol(), "", None));
+        assert!(test(parse_symbol(), "~()", Some(symbol!([]))));
+        assert!(test(parse_symbol(), ".", None));
+        assert!(test(parse_symbol(), "..", Some(symbol!([""]))));
+        assert!(test(parse_symbol(), "foo", Some(symbol!(["foo"]))));
+        assert!(test(parse_symbol(), "|foo|", Some(symbol!(["foo"]))));
+        assert!(test(
+            parse_symbol(),
+            "|foo|.|bar|",
+            Some(symbol!(["foo", "bar"]))
+        ));
+        assert!(test(
+            parse_symbol(),
+            ".|foo|.|bar|",
+            Some(symbol!(["foo", "bar"]))
+        ));
+        assert!(test(parse_symbol(), ".foo", Some(symbol!(["foo"]))));
```

Comment:
@johnchandlerburnham If the goal is to implement the agreed 'absolute path' meaning of initial `.`, let's add a test here:
```
        assert!(test(parse_symbol(), ".cons", Some(symbol!(["cons"]))));
```